cache/blobs.go

```diff
@@ -105,7 +121,7 @@ func computeBlobChain(ctx context.Context, sr *immutableRef, createIfNeeded bool
 					lowerRef = sr.layerParent
 				}
 				var lower []mount.Mount
-				if lowerRef != nil {
+				if !isOverlaybd && lowerRef != nil {
```

Comment:
The code is becoming quite complex with these conditions added. Quite some branches are excluded if `!isOverlaybd`, and stripping away that code shows that there's actually some of the other branches that are "dead code" in that case.

Perhaps out of scope for this PR, but I wonder if we need to branch earlier if `isOverlaybd` is true, instead of all these "micro-conditions".

I did a very quick / naive remove of all the `!isOverlaybd` paths, to get a "rough" idea what parts were unused, which gave me; 

<details>

```patch
diff --git a/cache/blobs.go b/cache/blobs.go
index a8af0a96f..b3abd863b 100644
--- a/cache/blobs.go
+++ b/cache/blobs.go
@@ -7,7 +7,6 @@ import (
 	"io"
 	"os"
 	"path"
-	"strconv"
 
 	obdlabel "github.com/containerd/accelerated-container-image/pkg/label"
 	obdcmd "github.com/containerd/accelerated-container-image/pkg/utils"
@@ -113,97 +112,11 @@ func computeBlobChain(ctx context.Context, sr *immutableRef, createIfNeeded bool
 					}
 					mediaType = ocispecs.MediaTypeImageLayer
 				}
-				var lowerRef *immutableRef
-				switch sr.kind() {
-				case Diff:
-					lowerRef = sr.diffParents.lower
-				case Layer:
-					lowerRef = sr.layerParent
-				}
 				var lower []mount.Mount
-				if !isOverlaybd && lowerRef != nil {
-					m, err := lowerRef.Mount(ctx, true, s)
-					if err != nil {
-						return nil, err
-					}
-					var release func() error
-					lower, release, err = m.Mount()
-					if err != nil {
-						return nil, err
-					}
-					if release != nil {
-						defer release()
-					}
-				}
-
-				var upperRef *immutableRef
-				switch sr.kind() {
-				case Diff:
-					upperRef = sr.diffParents.upper
-				default:
-					upperRef = sr
-				}
 				var upper []mount.Mount
-				if !isOverlaybd && upperRef != nil {
-					m, err := upperRef.Mount(ctx, true, s)
-					if err != nil {
-						return nil, err
-					}
-					var release func() error
-					upper, release, err = m.Mount()
-					if err != nil {
-						return nil, err
-					}
-					if release != nil {
-						defer release()
-					}
-				}
-
 				var desc ocispecs.Descriptor
 				var err error
 
-				// Determine differ and error/log handling according to the platform, envvar and the snapshotter.
-				var enableOverlay, fallback, logWarnOnErr bool
-				if forceOvlStr := os.Getenv("BUILDKIT_DEBUG_FORCE_OVERLAY_DIFF"); forceOvlStr != "" && sr.kind() != Diff {
-					enableOverlay, err = strconv.ParseBool(forceOvlStr)
-					if err != nil {
-						return nil, errors.Wrapf(err, "invalid boolean in BUILDKIT_DEBUG_FORCE_OVERLAY_DIFF")
-					}
-					fallback = false // prohibit fallback on debug
-				} else if !isTypeWindows(sr) {
-					enableOverlay, fallback = true, true
-					switch sr.cm.Snapshotter.Name() {
-					case "overlayfs", "stargz":
-						// overlayfs-based snapshotters should support overlay diff except when running an arbitrary diff
-						// (in which case lower and upper may differ by more than one layer), so print warn log on unexpected
-						// failure.
-						logWarnOnErr = sr.kind() != Diff
-					case "fuse-overlayfs", "native":
-						// not supported with fuse-overlayfs snapshotter which doesn't provide overlayfs mounts.
-						// TODO: add support for fuse-overlayfs
-						enableOverlay = false
-					}
-				}
-				if !isOverlaybd && enableOverlay {
-					computed, ok, err := sr.tryComputeOverlayBlob(ctx, lower, upper, mediaType, sr.ID(), compressorFunc)
-					if !ok || err != nil {
-						if !fallback {
-							if !ok {
-								return nil, errors.Errorf("overlay mounts not detected (lower=%+v,upper=%+v)", lower, upper)
-							}
-							if err != nil {
-								return nil, errors.Wrapf(err, "failed to compute overlay diff")
-							}
-						}
-						if logWarnOnErr {
-							bklog.G(ctx).Warnf("failed to compute blob by overlay differ (ok=%v): %v", ok, err)
-						}
-					}
-					if ok {
-						desc = computed
-					}
-				}
-
 				if isOverlaybd {
 					if err := commitOverlaybd(ctx, sr, &desc); err != nil {
 						return nil, err
```

</details>

Reply:
This code has been modified, because `desc` will be overwritten in `commitOverlaybd`, so there is no effect if `!isOverlaybd` is not added.